nvflare/fuel/utils/wfconf.py

```diff
@@ -58,14 +60,96 @@ def process_element(self, node: Node):
                 parent_element[node.key] = element
 
     def substitute(self, element: str):
-        a = re.split("{|}", element)
-        if len(a) == 3 and a[0] == "" and a[2] == "":
-            element = self.vars.get(a[1], None)
+        original_value = element
+
+        # Check for single item ref
+        # Single ref is resolved to an object that is derived from the item definition.
+        # If the item def also contains refs, all such refs will also be resolved.
+        # If the item def contains local vars, they are also resolved with the values from the ref.
+        # There are two kinds of Single Item Ref:
+        # - Simple ref that contains a single var name: {var_name}
+        # - Invoke a definition that contains local vars: {@var_name:p1=v1:p2=v2:...}
+        # The "@var_name" is a def that contains local vars p1, p2, ...
+        # When invoking such def, local var values could also be refs: {@var_name:p1={varp_name}}
+        is_single_ref = False
+        exp = element.strip()
+        if exp.startswith("{@") and exp.endswith("}"):
+            # this is a ref with local vars
+            is_single_ref = True
+            exp = exp[1 : len(exp) - 1]
+        else:
+            a = re.split("{|}", exp)
+            if len(a) == 3 and a[0] == "" and a[2] == "":
+                is_single_ref = True
+                exp = a[1]
+
+        if is_single_ref:
+            parts = exp.split(":")
+            var_name = parts[0]
+            params = []
+            for i, p in enumerate(parts):
+                if i > 0:
+                    params.append(p)
+
+            if params:
+                # the var_name must reference a dict
+                local_vars = parse_vars(params)
+                item = self.vars.get(var_name)
+                if item:
+                    if isinstance(item, dict):
+                        # scan the item to resolve var refs
+                        new_item = copy.deepcopy(item)
+                        scanner = JsonScanner(new_item)
+                        new_vars = copy.copy(self.vars)
```

Comment:
Do we need deepcopy for this as well?

Reply:
No, because "vars" are not updated. The reason that we need "new_vars" (and cannot use self.vars directly here) is because we update it with "local_vars".